app/components/primer/select_menu_component.rb

```diff
@@ -0,0 +1,550 @@
+    DEFAULT_BLANKSLATE = false
+    DEFAULT_ALIGN_RIGHT = false
+
+    with_slot :summary, class_name: "Summary"
+    with_slot :header, class_name: "Header"
+    with_slot :item, class_name: "Item", collection: true
+    with_slot :tab, class_name: "Tab", collection: true
+    with_slot :filter, class_name: "Filter"
+    with_slot :footer, class_name: "Footer"
+
+    attr_reader :message
+
+    #
+    # @example 193|Basic example|
+    #   <%= render Primer::SelectMenuComponent.new do |component| %>
+    #     <%= component.slot(:summary) do %>
+    #       Choose an option
+    #     <% end %>
+    #     <%= component.slot(:header) do %>
+    #       My menu
+    #     <% end %>
+    #     <%= component.slot(:item, selected: true, icon: "check") do %>
+    #       Item 1
+    #     <% end %>
+    #     <%= component.slot(:item, icon: "check") do %>
+    #       Item 2
+    #     <% end %>
+    #     <%= component.slot(:item, icon: "check") do %>
+    #       Item 3
+    #     <% end %>
+    #   <% end %>
+    #
+    # @example 193|Close button|Include a button to close the menu:
+    #   <%= render Primer::SelectMenuComponent.new do |component| %>
+    #     <%= component.slot(:summary) do %>
+    #       Choose an option
+    #     <% end %>
+    #     <%= component.slot(:header, closeable: true) do %>
+    #       My menu
+    #     <% end %>
+    #     <%= component.slot(:item) do %>
+    #       Item 1
+    #     <% end %>
+    #     <%= component.slot(:item) do %>
+    #       Item 2
+    #     <% end %>
+    #     <%= component.slot(:item) do %>
+    #       Item 3
+    #     <% end %>
+    #   <% end %>
+    #
+    # @example 242|Filter|If the list is expected to get long, consider adding a filter input. On mobile devices it will add a fixed height and anchor the select menu to the top of the screen. This makes sure the filter input stays at the same position while typing.
+    #   <%= render Primer::SelectMenuComponent.new do |component| %>
+    #     <%= component.slot(:summary) do %>
+    #       Choose an option
+    #     <% end %>
+    #     <%= component.slot(:header) do %>
+    #       My menu
+    #     <% end %>
+    #     <%= component.slot(:filter) %>
+    #     <%= component.slot(:item) do %>
+    #       Item 1
+    #     <% end %>
+    #     <%= component.slot(:item) do %>
+    #       Item 2
+    #     <% end %>
+    #     <%= component.slot(:item) do %>
+    #       Item 3
+    #     <% end %>
+    #   <% end %>
+    #
+    # @example 169|Tabs|Sometimes you need two or more lists of items in your select menu, e.g. branches and tags.
+    #   <%= render Primer::SelectMenuComponent.new do |component| %>
+    #     <%= component.slot(:summary, title: "Pick an item") do %>
+    #       Choose an option
+    #       <span class="dropdown-caret"></span>
+    #     <% end %>
+    #     <%= component.slot(:tab, selected: true) do %>
+    #       Tab 1
+    #     <% end %>
+    #     <%= component.slot(:tab) do %>
```

Comment:
How do you feel about declaring `tab`s and `item`s as siblings? It feels a little weird to me that items are not explicitly expressed as children of tabs.

@BlakeWilliams is this a case where we could nest one inside the other with Slots V2?

Reply:
I felt a bit weird having them as siblings, yeah. I'd love to define the items in the tabs they're meant to show in!

Reply:
This is really interesting!

I think nesting slots here *should* work.


e.g.

```ruby
class TabsComponent < ViewComponent::Base
  include ViewComponent::SlotableV2

  renders_many :tabs, MyTabComponent # implementation omitted
end
```

```ruby
class SelectMenuComponent < ViewComponent::Base
  include ViewComponent::SlotableV2

  renders_one :tab_bar, TabsComponent
end
```

```erb
<%= render SelectMenuComponent.new do |c| %>
  <%= c.tab_bar do |tab_bar| %>
    <%= tab_bar.tab { "one" }%>
    <%= tab_bar.tab { "two" }%>
  <% end %>
  
  <%= c.tab_bar { "three" } %>
<% end %>
```

If that *doesn't* work, I think we'd want to modify the implementation to ensure it does. That being said, I'm like 90% sure it should work.

Reply:
```ruby
# implementation omitted
```

Lol, I would also have accepted "implementation is left as an exercise for the student." 😄

Reply:
Thank you, Blake! Gonna try this now.

Reply:
I need to define both the clickable part of the tab as well as the items that appear in its body. Your `<%= tab_bar.tab { "one" }%>` example looks like what would be in the clickable part. I'm wondering what a good interface would look like to convey both parts. Maybe:

```erb
<%= render SelectMenuComponent.new do |c| %>
  <%= c.tab do |tab_component| %>
    <%= tab_component.button do %>
      [icon] Tab 1
    <% end %>
    <%= c.slot(:item) do %>
      first item in tab 1
    <% end %>
    <%= c.slot(:item) do %>
      second item in tab 1
    <% end %>
  <% end %>

  <%= c.tab do |tab_component| %>
    <%= tab_component.button do %>
      [icon] Tab 2
    <% end %>
    <%= c.slot(:item) do %>
      first item in tab 2
    <% end %>
    <%= c.slot(:item) do %>
      second item in tab 2
    <% end %>
  <% end %>
<% end %>
```

Reply:
Attempt in process in https://github.com/primer/view_components/pull/142/commits/f900cdc8b7da43b65fb91d9c72f89ae761786040. 😅